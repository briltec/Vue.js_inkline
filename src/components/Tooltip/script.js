import { uid } from '@inkline/inkline/helpers';

import AttributesProviderMixin from '@inkline/inkline/mixins/components/providers/AttributesProviderMixin';
import ClassesProviderMixin from '@inkline/inkline/mixins/components/providers/ClassesProviderMixin';
import EmitProviderMixin from '@inkline/inkline/mixins/components/providers/EmitProviderMixin';
import PopupProviderMixin from '@inkline/inkline/mixins/components/providers/PopupProviderMixin';
import PopupControlsProviderMixin from '@inkline/inkline/mixins/components/providers/PopupControlsProviderMixin';

import EmitFocusMethodMixin from '@inkline/inkline/mixins/components/methods/EmitFocusMethodMixin';

import SizePropertyMixin from '@inkline/inkline/mixins/components/properties/SizePropertyMixin';
import VariantPropertyMixin from '@inkline/inkline/mixins/components/properties/VariantPropertyMixin';
import DisabledPropertyMixin from '@inkline/inkline/mixins/components/properties/DisabledPropertyMixin';

import ClickOutside from '@inkline/inkline/directives/click-outside';

export default {
    name: 'ITooltip',
    mixins: [
        AttributesProviderMixin,
        ClassesProviderMixin,
        EmitProviderMixin,
        PopupProviderMixin,
        PopupControlsProviderMixin,

        EmitFocusMethodMixin,

        SizePropertyMixin,
        VariantPropertyMixin,
        DisabledPropertyMixin,
    ],
    directives: {
        ClickOutside
    },
    props: {
        trigger: {
            type: String,
            default: 'hover'
        },
        placement: {
            type: String,
            default: 'top'
        },
        arrow: {
            type: Boolean,
            default: true
        },
        arrowOffset: {
            type: Number,
            default: 0
        }
    },
    data() {
        const basename = 'tooltip';

        return {
            id: this.$attrs.id || uid(basename),
            basename
        };
    },
    watch: {
        'placement': {
            immediate: true,
                handler(value) {
                this.currentPlacement = value;
            }
        }
    },
    created() {
        this.$on('updatePopper', () => {
            if (this.visible) this.updatePopper();
        });
    },
    mounted() {
        this.referenceElement = this.$el;
        this.currentPlacement = this.placement;
    }
};
