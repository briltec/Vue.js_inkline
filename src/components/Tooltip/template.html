<div class="tooltip-wrapper">
    <slot />

    <transition name="zoom-in-top-transition" @after-leave="doDestroy">
        <div class="tooltip" ref="tooltip" :class="classes['*']" v-show="visible">
            <span class="arrow" v-if="arrow"></span>
            <slot name="body"></slot>
        </div>
    </transition>
</div>