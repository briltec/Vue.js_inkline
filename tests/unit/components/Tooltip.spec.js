import { shallowMount } from '@vue/test-utils';
import Tooltip from '@inkline/inkline/components/Tooltip';

describe('Components', () => {
    describe('Popover', () => {
        let wrapper;

        beforeEach(() => {
            wrapper = shallowMount(Tooltip, {
                propsData: {
                    id: 'tooltip'
                },
                methods: {
                    created: Tooltip.created,
                    mounted: Tooltip.mounted
                },
                slots: {
                    default: ['<div/>']
                }
            });
        });

        it('should be named correctly', () => {
            expect(Tooltip.name).toEqual('ITooltip');
        });

        it('should render correctly', () => {
            expect(wrapper.html()).toMatchSnapshot();
        });

        describe('props', () => {
            describe('trigger', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.trigger).toBeDefined();
                    expect(wrapper.vm.trigger).toEqual('hover');
                });
            });

            describe('placement', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.placement).toBeDefined();
                    expect(wrapper.vm.placement).toEqual('top');
                });
            });

            describe('arrow', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.arrow).toBeDefined();
                    expect(wrapper.vm.arrow).toEqual(true);
                });
            });

            describe('arrowOffset', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.arrowOffset).toBeDefined();
                    expect(wrapper.vm.arrowOffset).toEqual(0);
                });
            });
        });

        describe('data()', () => {
            describe('basename', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.basename).toBeDefined();
                    expect(wrapper.vm.basename).toEqual('tooltip');
                });
            });

            describe('id', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.id).toBeDefined();
                    expect(wrapper.vm.id).toEqual('tooltip');
                });

                it('should be randomly generated if not defined', () => {
                    wrapper = shallowMount(Tooltip, {
                        slots: {
                            default: ['<div/>']
                        }
                    });

                    expect(wrapper.vm.id).toMatch(/^tooltip-\w+/);
                });
            });
        });

        describe('created()', () => {
            it('should add updatePopper event listener', () => {
                const spy = jest.spyOn(wrapper.vm, '$on');

                wrapper.vm.created();

                expect(spy).toHaveBeenCalled();
                expect(spy).toHaveBeenCalledWith('updatePopper', expect.any(Function));
            });
        });

        describe('mounted()', () => {
            it('should set reference element and current placement', () => {
                wrapper.vm.$el = 'new';

                wrapper.vm.mounted();

                expect(wrapper.vm.referenceElement).toEqual('new');
            });
        });

        describe('watch', () => {
            describe('placement', () => {
                it('should set current placement to updated placement value', () => {
                    expect(wrapper.vm.currentPlacement).toEqual('top');
                    wrapper.setProps({ placement: 'new' });
                    expect(wrapper.vm.currentPlacement).toEqual('new');
                });
            });
        });
    });
});
