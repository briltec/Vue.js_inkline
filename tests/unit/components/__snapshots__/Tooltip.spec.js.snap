// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Popover should render correctly 1`] = `
<div class="tooltip-wrapper" id="tooltip">
  <div aria-haspopup="tooltip" aria-controls="tooltip"></div>
  <div class="tooltip" style="display: none;" name="zoom-in-top-transition" id="tooltip"><span class="arrow"></span> </div>
</div>
`;
